subcommands/targets/offline-update.go

```diff
@@ -0,0 +1,279 @@
+package targets
+
+import (
+	"archive/tar"
+	"compress/bzip2"
+	"encoding/json"
+	"fmt"
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"io"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	factory string
+	prod    bool
+	tag     string
+)
+
+type (
+	targetInfo struct {
+		version       int
+		ostreeVersion int
+		hardwareID    string
+		buildTag      string
+	}
+)
+
+func init() {
+	offlineUpdateCmd := &cobra.Command{
+		Use:   "offline-update <target-name> <dst> --tag <tag> [--prod]",
+		Short: "Download Target content for an offline update",
+		Run:   doOfflineUpdate,
+		Args:  cobra.ExactArgs(2),
+	}
+	cmd.AddCommand(offlineUpdateCmd)
+	offlineUpdateCmd.Flags().BoolVarP(&prod, "prod", "", false,
+		"Instruct to fetch content of production Target")
+	offlineUpdateCmd.Flags().StringVarP(&tag, "tag", "", "",
+		"Target tag")
+}
+
+func doOfflineUpdate(cmd *cobra.Command, args []string) {
+	factory = viper.GetString("factory")
+	targetName := args[0]
+	dstDir := args[1]
+
+	if len(tag) == 0 {
+		fmt.Fprintln(os.Stderr, "Mandatory flag `tag` is not specified")
+		os.Exit(1)
+	}
+
+	fmt.Printf("Checking whether Target exists; target: %s, tag: %s, production: %v\n", targetName, tag, prod)
+	ok, err := checkIfTargetExists(targetName)
+	subcommands.DieNotNil(err, "Failed to check whether Target exists:")
+	if !ok {
+		fmt.Fprintf(os.Stderr, "The specified Target has not been found; target: %s, tag: %s, production: %v\n", targetName, tag, prod)
+		os.Exit(1)
```

Comment:
again - volodya converted all our code to exit prematurely with `subcommands.DieNotNil`. Let's stay with that for now. You could probably just have `checkIfTargetExists` return just an error and skip the OK. Or just call it `assertTargetExists` and die there.

Reply:
`checkIfTargetExists` fetches the tagged condensed Target, what a device would fetch, `getTargetInfo` fetches the original CI target in order to get the correct branch/tag.

Reply:
> again - volodya converted all our code to exit prematurely with subcommands.DieNotNil

Makes sense to some extend.

Reply:
i'm still not seeing why you can't do something like get the whole list of targets and return the `client.CustomTarget` for it rather than "ok" and pass that to `getTargetInfo`. I don't see what you are gaining by two round-trips to the server that can't be done in one trip?

Reply:
The targets downloaded for offline update do not contain the target custom required here (with original CI tag), so there is no two requests to the same handler. The TUF targets are downloaded from the repo API, while the target custom details are obtained from the `/ota/factories/<factory>/targets/target` call.

Reply:
exactly, e.g.

```
curl -s -H "OSF-TOKEN: `cat $FIO_TOKEN`" "${FIO_API}/repo/$FACTORY/api/v1/user_repo/targets.json?tag=beta&production=1" | jq '.signed.targets."raspberrypi4-64-lmp-1452"."custom"."tags"'
[
  "beta"
]
[mike@t14s ota-lite]$ curl -s -H "OSF-TOKEN: `cat $FIO_TOKEN`" "${FIO_API}/repo/$FACTORY/api/v1/user_repo/targets.json?tag=beta" | jq '.signed.targets."raspberrypi4-64-lmp-1452"."custom"."tags"'
[
  "master",
  "beta"
]

```

And making `if !prod` to optimize leads to more source code and less readability, and we think that the prod use-case will be prevailing anyway.